### DISCUSSION TITLE
Progress with pyMC

### DISCUSSION THREAD

**Discussion by shakasaki** [gempy-project/gempy#582] [Q&A] [answered]
**What is your question?**

Hello - I am wondering what is the progress with integrating Gempy and pyMC3? Is this still continuing despite the Theano discontinuation?

The current [tutorial](https://www.gempy.org/documentation) on probabilistic modeling is a bit outdated, specifically regarding PyMC3. Is there any more information to share on this?

What I'm trying to do is to combine uncertainty in point locations and their orientations, and based on that compute surfaces from which I later want to extract topological properties. My models are "simple" in that they do not include faults. 

If anyone is working on this and is interested to hear more, i'd be happy to share!

Thanks for the feedback

**Comment by LukasMosser** [comment]
Follow up question on this: 
Since PyMC will most likely move to support of a jax-based theano replacement backend, is there any plan to move to that once pymc has enabled a jax backend?

**Comment by Leguark** [comment, marked as answer]
Thank you for the question, it is quite interesting times with so many news and technologies coming up every second week.

As @shakasaki  mentioned Theano was discontinued two years ago and slowly but steady is (was maybe) getting more painful to use. Specially with the support for RTX cards. 

For that reason we decided a year ago to rewrite the kernel in tensor flow, luckily because I expected that nowadays nothing is long lasting, I am trying to rewrite (https://github.com/Leguark/gempy_engine) the kernel  of gempy agnostic (to some extent) of any specific backend  - helped by the fact that most of this libraries are moving to have a numpy api or close to it. So far my plan is to make it compatible with:
- Numpy
- Tensor Flow (at least eager)
- Jax
- Pykeops

Now, as @LukasMosser said, while I was doing the refactoring the PyMC group came and said that they are working theano to use jax as backend. What this means is that without touching anything of gempy suddenly it could start working using jax simply by installing theano from a different repo. I did not try this yet - and I expect not to work yet since we are using quite specific theano functions - but this has the potential to solve many of the issues in the short term.

What does all of this mean to the future of gempy:
- Short-mid term:  hopefully the theano-jax implementation works fine and we can keep using pymc3 
- Mid-Long term: the refactoring is going to happen in any case and in principle will come online in 2021 with a bunch of extra features. If we interpolate directly in JAX/TF probably we will shift to TF probability for Bayesian networks.

I do not if I answered your questions but at least I hope now it is more transparent what we are hoping for in the future

**Comment by shakasaki** [comment]
Hi @Leguark and thanks for clearing this out! It seems like a long and winding road, but I look forward to see the outcome.

Would this [publication](https://gmd.copernicus.org/preprints/gmd-2020-136/gmd-2020-136.pdf) be the most relevant and recent example of using gempy/theano in a probabilistic framework? (short-mid term solution)

**Reply by Leguark** [reply]
Sorry for the very late response. If I recall properly in that paper we did not use pymc3 but the main author wrote an implementation of MC rejection sampling. The most recent article we wrote using pymc3 and actually NUTS was https://www.mdpi.com/2076-3263/11/4/150
